packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -0,0 +1,444 @@
+const double _kInspectButtonMargin = 10.0;
+
+/// Interpret pointer up events within with this margin as indicating the
+/// pointer is moving off the device.
+const double _kOffScreenMargin = 1.0;
+
+// TODO(jacobr): merge with similar code in tooltip.dart.
+class _TooltipPositionCalculator {
+  _TooltipPositionCalculator({
+    this.target,
+    this.verticalOffset,
+    this.preferBelow,
+    this.size,
+    this.childSize,
+  }) {
+    // VERTICAL DIRECTION
+    final bool fitsBelow = target.dy + verticalOffset + childSize.height <=
+        size.height - _kScreenEdgeMargin;
+    final bool fitsAbove =
+        target.dy - verticalOffset - childSize.height >= _kScreenEdgeMargin;
+    _tooltipBelow =
+        preferBelow ? fitsBelow || !fitsAbove : !(fitsAbove || !fitsBelow);
+    double y;
+    if (_tooltipBelow)
+      y = math.min(
+          target.dy + verticalOffset, size.height - _kScreenEdgeMargin);
+    else
+      y = math.max(
+          target.dy - verticalOffset - childSize.height, _kScreenEdgeMargin);
+    // HORIZONTAL DIRECTION
+    final double normalizedTargetX =
+        target.dx.clamp(_kScreenEdgeMargin, size.width - _kScreenEdgeMargin);
+    double x;
+    if (normalizedTargetX > size.width - _kScreenEdgeMargin - childSize.width) {
+      // Make room for content to the right.
+      x = size.width - _kScreenEdgeMargin - childSize.width;
+    } else {
+      x = normalizedTargetX;
+    }
+    _childPosition = new Offset(x, y);
+  }
+
+  final Offset target;
+  final double verticalOffset;
+  final bool preferBelow;
+  final Size size;
+  final Size childSize;
+
+  Offset _childPosition;
+  bool _tooltipBelow;
+
+  bool get tooltipBelow => _tooltipBelow;
+  Offset get childPosition => _childPosition;
+}
+
+// FOR DISCUSSION: this is clearly not quite right: should I even even bother
+// listening for changes? This code is copied from semantics_debugger where
+// it made sense.
+class _WidgetClient extends ChangeNotifier {
+  _WidgetClient(PipelineOwner pipelineOwner) {
+    _semanticsHandle =
+        pipelineOwner.ensureSemantics(listener: _didUpdateSemantics);
+  }
+
+  SemanticsHandle _semanticsHandle;
+
+  @override
+  void dispose() {
+    _semanticsHandle.dispose();
+    _semanticsHandle = null;
+    super.dispose();
+  }
+
+  int generation = 0;
+
+  void _didUpdateSemantics() {
+    generation += 1;
+    notifyListeners();
+  }
+}
+
+const TextStyle _messageStyle = const TextStyle(
+    color: const Color(0xFFFFFFFF), fontSize: 10.0, height: 1.2);
+final int _kMaxTooltipLines = 5;
+const Color _kTooltipBackgroundColor = const Color.fromARGB(230, 60, 60, 60);
+
+void _paintDescription(Canvas canvas, RenderBox render, Offset target,
+    double verticalOffset, Size size, Rect targetRect) {
+  // TODO(jacobr): craft a better description message.
+  final String message = '$render\n\n'
+      '${render.debugCreator}';
+  canvas.save();
+  final TextPainter textPainter = new TextPainter()
```

Comment:
Now that I'm caching the output to a layer and only re-rendering on state change, does attempting to cache the TextPainter really matter? Any time the message or the width of the selected render object changes the painter would need to be updated anyway.